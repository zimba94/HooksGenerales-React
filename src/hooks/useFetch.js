import {useState, useEffect, useRef} from 'react';

export const useFetch = (url) => {

    const isMounted = useRef(true);   
    const [state, setstate] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {

        setstate({data: null, loading: true, error: null});
        
        fetch(url)
            .then(resp => resp.json())
            .then(data => {

                if (isMounted.current) {
                    setstate({
                        loading: false,
                        error: false,
                        data
                    });
                }
            })
            .catch(()=>{
                setstate({
                    data:null,
                    loading: false,
                    error: 'No se pudo cargar la info'
                })
            });
    }, [url]);

    return state;
}
