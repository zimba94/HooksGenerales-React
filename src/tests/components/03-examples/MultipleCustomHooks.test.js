import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultipleCustomHooks />', () => {

    beforeEach( () => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
    });

    test('Debe mostrarse correctamente', () => {
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        });
        const wrapper = shallow(< MultipleCustomHooks />);
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe mostrar la informaciÃ³n', () => {
        useFetch.mockReturnValue({
            data:[{
                author: 'Alan',
                quote: 'Hola mundo',
            }],
            loading:false,
            error:null
        });
        const wrapper = shallow(< MultipleCustomHooks />);
        expect( wrapper.find('.alert').exists()).toBe(false);
        expect( wrapper.find('.mb-3').text().trim()).toBe('Hola mundo');
        expect( wrapper.find('footer').text().trim()).toBe('Alan');

        console.log(wrapper.html());
    });
    

});
