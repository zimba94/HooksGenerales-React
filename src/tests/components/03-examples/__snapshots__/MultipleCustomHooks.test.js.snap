// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Hooks!!!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
